### ISSUE TITLE
MySQL `upsert()`: `Internal error: Attempted to serialize empty result.`

### ISSUE THREAD

**Issue by jde** [prisma/prisma#20042]
Hi Prisma Team! My Prisma Client just crashed.

~~I'm not able to reproduce this error. So far I've gotten this exactly once. When I try the exact same request again the query executes as expected. I hope this helps track down an issue. Otherwise feel free to close.~~

This is happening every time I execute the upsert with an id in the where clause, aka, every time I update.

THANK YOU FOR AN AMAZING PROJECT!

Dave

This is the report:
## Versions

| Name            | Version            |
|-----------------|--------------------|
| Node            | v16.13.1           | 
| OS              | darwin             |
| Prisma Client   | 4.14.1             |
| Query Engine    | d9a4c5988f480fa576d43970d5a23641aa77bc9c|
| Database        | mysql              |

## Logs
```
prisma:tryLoadEnv Environment variables loaded from /Users/jde/dev/github/rolldeeep/admin/.env
prisma:client checkPlatformCaching:postinstall false
prisma:client checkPlatformCaching:ciName 
prisma:tryLoadEnv Environment variables loaded from /Users/jde/dev/github/rolldeeep/admin/.env
prisma:client dirname /Users/jde/dev/github/rolldeeep/admin/node_modules/.prisma/client
prisma:client relativePath ../../../prisma
prisma:client cwd /Users/jde/dev/github/rolldeeep/admin/prisma
prisma:client protocol graphql
prisma:client clientVersion 4.14.1
prisma:client clientEngineType library
prisma:client:libraryEngine internalSetup
prisma:client:libraryEngine:loader Searching for Query Engine Library in /Users/jde/dev/github/rolldeeep/admin/node_modules/.prisma/client
prisma:client:libraryEngine:loader loadEngine using /Users/jde/dev/github/rolldeeep/admin/node_modules/.prisma/client/libquery_engine-darwin.dylib.node
prisma:client:libraryEngine library starting
prisma:client:libraryEngine library started
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
prisma:client:libraryEngine sending request, this.libraryStarted: true
```

## Client Snippet
```ts
export async function upsertVendor(data) {

  const d = {
    name: data.name,
    contact: data.contact,
    email: data.email || '',
    phone: data.phone || '',
    description: data.description || '',
    image: data.image || '',
    url: data.url || '',
    notes: data.notes || '',
  }

  const upsert = {
    where: { id: data.id || '' },
    update: d,
    create: d,
  }

  const vendor = await prisma.vendor.upsert(upsert)
  return vendor
}
```

## Schema
```prisma
model Vendor {
  id String @id @default(cuid())

  name          String  @unique
  contact       String
  email         String
  emailVerified Boolean @default(false)
  phone         String
  address       String? @db.Text
  description   String  @default("")
  image         String?
  url           String? // External url for more information about the vendor.

  notes String @default("") @db.Text // Internally facing notes

  relationshipType String @default("None") // Partner, Affiliate, Referral, None

  // Is this a vendor record for a property?
  property   Property? @relation(fields: [propertyId], references: [id], onUpdate: NoAction)
  propertyId String?   @unique

  users      VendorsOnUsers[] // Users who can manage this vendor record.
  services   Service[] // Services they provide.
  properties VendorsOnProperties[] // Properties they service.
  tripLogs   TripLog[] // Logs left involving them.
  categories Category[]

  createdAt DateTime  @default(now())
  updatedAt DateTime? @updatedAt
  deletedAt DateTime?

  @@index([id])
  @@index([email])
  @@map("rd_vendor")
}
```

## Prisma Engine Query
```
{
  where: { id: 'cljn25lsi0000lx4mhxnljhsr' },
  update: {
    name: 'abcabc',
    contact: 'xyzxyz',
    email: 'some@email.co',
    phone: '1231231234',
    description: '',
    image: '',
    url: '',
    notes: ''
  },
  create: {
    name: 'abcabc',
    contact: 'xyzxyz',
    email: 'some@email.co',
    phone: '1231231234',
    description: '',
    image: '',
    url: '',
    notes: ''
  }
}
```

**Comment by jde** [comment]
I'm now seeing this error on an ongoing basis. So far it is always when I'm updating through the upsert. I still haven't been able to find the specific sequence of events that leads to this. I'll continue to scrutinize it.

This is a blocker for my application as it breaks functionality.

The error message is:
```
thread 'tokio-runtime-worker' panicked at 'Internal error: Attempted to serialize empty result.', query-engine/core/src/response_ir/ir_serializer.rs:58:40
PrismaClientRustPanicError:
Invalid `prisma.vendor.upsert()` invocation:
```

**Comment by jde** [comment]
I'm seeing in other issues that this bug seems to have been introduced somewhere after v4.5.0, and that there is currently no resolution.

https://github.com/prisma/prisma/issues?q=is%3Aissue+is%3Aopen+Attempted+to+serialize+empty+result

What is the recommended course of action? Should we pin our version at v4.5.0 until a fix is merged?

**Comment by jde** [comment]
I have temporarily gotten around this by reverting to v4.5.0 and using the deprecated referentialIntegrity feature. Hoping this can be resolved in a future patch!

https://github.com/prisma/prisma/issues/17279#issuecomment-1618899569

**Comment by janpio** [comment]
A full, minimal reproduction of the problem in a GitHub repository would be amazing for us to simplify reproducing and understanding the problem. Can you maybe invest the time to create one?

**Comment by jde** [comment]
Sorry to have let this sit for so long. I've been pulled in a number of directions. I'll see if I can recreate this in a minimal form asap.

**Comment by homerjam** [comment]
Possibly related: #17305, #18285, #19868

I was having similar issues after upgrading from 3.15.2, my database is on Planetscale. I did two things which I think helped over come the problems.

### Removed referential action `onUpdate: NoAction` from `@relation()` 
I believe this was necessary due to how Prisma used to work around missing foreign key support in Planetscale, now I don't think I need it (we'll see, this is a side project)

### Removed the `id` property in the `where` clause from the `update` payload of the `upsert()` action
This wasn't needed in the payload (laziness on my part) but was causing the `violate the required relation` error initially, I think, and the subsequently causing the record not to be found.

Hopefully this might help someone!

**Comment by chadedrupt** [comment]
I had exactly the same issue.
I did not have the `id` used in the `where` in the `update` section of the `upsert`.
However, I did have some rogue, unnecessary `onUpdate: NoAction` not on the problem model itself, but in some deeper child `relations`.
Removing the unnecessary `onUpdate: NoAction` fixed it for me.

**Comment by janpio** [comment]
@jde @chadedrupt @homerjam @jde Are you all using `relationMode=prisma`?

**Comment by laplab** [comment]
This issue was fixed in [Prisma 5.13.0](https://github.com/prisma/prisma/releases/tag/5.13.0).
